### PR TITLE
openvpn: use hotplug.d

### PR DESCRIPTION
**Pull request by feckert**
Add to openvpn the good old openwrt hotplug mechanism.
There are also the following config option which could also use **hotplug.d**.
- route-up
- ipchange
- tls-verify
- client-connect
- learn-address
- auth-user-pass

### CODE REVIEW

File: package/network/services/openvpn/files/openvpn.init [hunk 1]
```diff
@@ -73,12 +73,30 @@ openvpn_add_instance() {
 	local dir="$2"
 	local conf="$3"
 
+	local up down
+
+	config_get up "$name" up
+	[ -z "$up" ] && {
```
Comment (jow-) [inline package/network/services/openvpn/files/openvpn.init]: I think that when an up or down command is already configured, you should wrap it instead of disabling the hotplug mechanism in this case.

**Comment by feckert** [comment]
@jow ping?

**Comment by jow-** [comment]
With wrapping it I mean that you first execute the new hotplug machinery, then you make the new hotplug machinery invoke whatever user-configured up script is there.

**Comment by feckert** [comment]
@jow I have update my pull request and added your suggestion to wrap user defined up/down scripts.

File: package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+# Wrap user defined scripts on up/down events
+. /lib/functions.sh
+config_load 'openvpn'
+LOG="logger -t openvpn/hotplug[$$] -p"
+
+[ "$ACTION" = "up" ] && {
+	config_get up "$INSTANCE" up
+	[ -z "$up" ] || {
+		if [ -f "$up" ]; then
+			/bin/sh "$up" \
```
Comment (jow-) [inline package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user]: Please use `[ -n "$up" ]` and `/bin/sh -c "$up" ...`.

The OpenVPN help states:
> --up cmd
> Shell  command  to run after successful TUN/TAP device open

So that seems to imply that both something like `echo 1` or `/usr/bin/example` is supposed to work.

File: package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+			$LOG warn "Configured $ACTION script \"$up\" for instance \"$INSTANCE\" not found"
+		fi
+	}
+}
+
+[ "$ACTION" = "down" ] && {
+	config_get down "$INSTANCE" down
+	[ -z "$down" ] || {
+		if [ -f "$down" ]; then
+			/bin/sh "$down" \
```
Comment (jow-) [inline package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user]: Likewise, `[ -n "$down" ]` and `/bin/sh -c ...`

File: package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user (hunk 2 above)
Comment (feckert) [reply package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user]: > Please use [ -n "$up" ] 

OK will change this

> So that seems to imply that both something like echo 1 or /usr/bin/example is supposed to work.

Yes this is right but I get the following error if I add `echo 1` to the **down** option of a openvpn uci config section. From my point of view it does not make sense to only add a command line string,  because you have to wrap all this quotation `' " `in the uci config. I think it would be better for the user to add their command into a script.
The advantage is he has all the possibilities a shell offers and he doesn't have to be annoyed with quotation marks.

```
config openvpn 'test'
     option down 'echo 1'
```
> Options error: the --down directive should have at most 1 parameter.  To pass a list of arguments as one of the parameters, try enclosing them in double quotes ("").
> Thu Apr  4 11:11:04 2019 daemon.warn openvpn(test)[20360]: Use --help for more information.

Comment (feckert) [reply package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user]: I have update the check but I think that using only a command and not a script is not the way to go. For the reasons given above.

Comment (jow-) [reply package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user]: It still requires the command to be a file. Command strings like `route add ... && route add ... && /etc/init.d/foo restart` are perfectly valid.

Comment (jow-) [reply package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user]: You also forgot to pass `-c` to `/bin/sh`. Without that, the shell will attempt to parse the argument as script, which will not work for ELF executables:

```
root@jj:~# /bin/sh /bin/date
/bin/date: line 2: syntax error: unterminated quoted string
root@jj:~# /bin/sh -c /bin/date
Wed Apr 24 11:03:30 CEST 2019
```

Comment (feckert) [reply package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user]: Sorry I do not get it.
If I set `--up "/usr/bin/logger hello1 && /usr/bin/logger hello"` only the first logger is called 

> openvpn(test)[3023]: /usr/bin/logger hello1 && /usr/bin/logger hello tun0 1500 1544 10.9.8.2 10.9.8.1 init
>  user.notice root: hello1 && /usr/bin/logger hello tun0 1500 1544 10.9.8.2 10.9.8.1 init

Comment (jow-) [reply package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user]: The following works in native configuration:

    up '/bin/sh -c "/usr/bin/logger test1 && /usr/bin/logger test2"'

The following works in UCI:

    option up "\"/bin/sh -c '/usr/bin/logger test1 && /usr/bin/logger test2'\""

Your check for `[ -f "$up" ]` breaks these use cases and forcing the command string through `/bin/sh` without `-c` breaks normal non-shellscript programs as well.

You may not make assumptions about the contents of `$up` and `$down` other than that they're  command vectors to be passed to `execve()`.  OpenVPN actually tokenizes the `up` or `down` values into lists while respecting quotes and escapes. So the example values above `/bin/sh -c "/usr/bin/logger test1 && /usr/bin/logger test2"` becomes `{ "/bin/sh", "-c", "/usr/bin/logger test1 && /usr/bin/logger test2" }` which is then passed to `execve()` as follows: `execve("/bin/sh", { "/bin/sh", "-c", "/usr/bin/logger test1 && /usr/bin/logger test2", "hello", "tun0", "1500", "1544", "10.9.8.2", "10.9.8.1", "init" }, envptr)`.

A simpler, more real word example would be something similar to `up "/usr/local/bin/my-custom-handler --action=init --instance=foo --"` which will get passed as `execve("/usr/local/bin/my-custom-handler", { "/usr/local/bin/my-custom-handler", "--action=init", "--instance=foo", "--", "hello", "tun0", "1500", "1544", "10.9.8.2", "10.9.8.1", "init" }, envptr)`.

So in short, testing for `[ -f "$up" ]` will break the ability to specify commands with arguments (which is supported by OpenVPN) and running through `/bin/sh` instead of `/bin/sh -c` will break all variants where the value refers to a path that is not a shell script.

Comment (feckert) [reply package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user]: @jow Thank you very much or the clarifying words. I have now added your requested changes. I have also remove **up**, **down** and **script_security** from the openvpn.options file. If I do not do this i will get the following erros during service start

> daemon.err openvpn(test)[2684]: Options error: the --down directive should have at most 1 parameter.  To pass a list of arguments as one of the parameters, try enclosing them in double quotes ("").

This is because the command will also get moved to the openvpn.config but I think that's wrong., because the options are now added during openvpn start. This must also be true for the remaining parameters **cd**, **status** and **syslog**. The option **config** is already not included in this file.

**Comment by feckert** [comment]
I would like to see the feature in Openwrt master.
Is there anything else I can do here?
It's been lying here for almost a year now.

This change would perfectly fit into the hotplug mechanism of openwrt.
With this feature you can for example set a LED when a tunnel has been built up or send an email.

**Comment by feckert** [comment]
ping @ynezz @dedeckeh 
Can someone **please** take a look at this?
All requested changes are done.

**Comment by bobafetthotmail** [comment]
A different approach: does does openvpn package really need to stay in main repo? Can it be moved with other VPNs in the packages repo?

Especially @ynezz as he is the one that wants to move non-core stuff to packages.

**Comment by jow-** [comment]
Gave this PR another try today and it unfortunately is not ready for merging yet.

Two main issues:

- If an instance is started from an `/etc/openvpn/*.conf` file which does not happen to be a valid uci identifier, the logic figuring out the `script_security` setting fails. In my case I do have a file `/etc/openvpn/gateway-madrid.conf` which results in an instance name of `gateway-madrid`. The causes `config_get()` to trip up:

```
+ config_get script_security gateway-madrid script_security
+ eval export -n -- 'script_security=${CONFIG_gateway-madrid_script_security:-${4}}'
+ export -n -- 'script_security=madrid_script_security:-'
+ '[' -z madrid_script_security:- ]
```
Due to that, the instance is launched with `--script-security madrid_script_security:-` and subsequently fails to start with: 

```
Tue Jun 16 21:51:54 2020 daemon.warn openvpn(gateway-madrid)[27112]: WARNING: External program may not be called unless '--script-security 2' or higher is enabled. See --help text or man page for detailed info.
Tue Jun 16 21:51:54 2020 daemon.err openvpn(gateway-madrid)[27112]: Linux ip link set failed: disallowed by script-security setting
Tue Jun 16 21:51:54 2020 daemon.notice openvpn(gateway-madrid)[27112]: Exiting due to fatal error
```

Second issue:
The `/etc/hotplug.d/openvpn/01-user` script does not honour `up` and `down` in native OpenVPN configuration files, so this change will completely break `--up` and `--down` for OpenVPN instances not using the uci format.

**Comment by jow-** [comment]
Another problem:

Since the hotplug-call is invoked without `exec` and using `env -i` the environment variables passed by the OpenVPN process (such as `$dev`, `$ifconfig_local` and so on) are lost and not available to either `/etc/hotplug.d/openvpn/01-user` nor the scripts invoked by it.

**Comment by jow-** [comment]
@feckert - the following change on top of your PR fixes things for me:
```diff
diff --git a/package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user b/package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user
index 9b06ed5ef7..82e25980a8 100644
--- a/package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user
+++ b/package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user
@@ -1,5 +1,18 @@
 #!/bin/sh
 
+get_option() {
+	local variable="$1"
+	local option="$2"
+
+	local value="$(sed -rne 's/^[ \t]*'"$option"'[ \t]+([^ \t]+)[ \t]*$/\1/p' "$config" | tail -n1)"
+	[ -n "$value" ] || value="$(sed -rne 's/^[ \t]*'"$option"'[ \t]+'"'([^']+)'"'[ \t]*$/\1/p' "$config" | tail -n1)"
+	[ -n "$value" ] || value="$(sed -rne 's/^[ \t]*'"$option"'[ \t]+"(([^"\\]|\\.)+)"[ \t]*$/\1/p' "$config" | tail -n1 | sed -re 's/\\(.)/\1/g')"
+	[ -n "$value" ] || return 1
+
+	export -n "$variable=$value"
+	return 0
+}
+
 [ -e "/etc/openvpn.user" ] && {
 	env -i ACTION="$ACTION" INSTANCE="$INSTANCE" \
 		/bin/sh \
@@ -9,22 +22,12 @@
 
 # Wrap user defined scripts on up/down events
 
-[ "$ACTION" = "up" ] && {
-	. /lib/functions.sh
-	config_load 'openvpn'
-	config_get up "$INSTANCE" up
-	[ -n "$up" ] &&  {
-		/bin/sh -c "$up $ACTION $INSTANCE $*"
-	}
-}
-
-[ "$ACTION" = "down" ] && {
-	. /lib/functions.sh
-	config_load 'openvpn'
-	config_get down "$INSTANCE" down
-	[ -n "$down" ] && {
-		/bin/sh -c "$down $ACTION $INSTANCE $*"
-	}
-}
+case "$ACTION" in
+	up|down)
+		if get_option command "$ACTION"; then
+			exec /bin/sh -c "$command $ACTION $INSTANCE $*"
+		fi
+	;;
+esac
 
 exit 0
diff --git a/package/network/services/openvpn/files/openvpn.init b/package/network/services/openvpn/files/openvpn.init
index ff8cdd12d8..72882550fb 100644
--- a/package/network/services/openvpn/files/openvpn.init
+++ b/package/network/services/openvpn/files/openvpn.init
@@ -73,13 +73,7 @@ openvpn_add_instance() {
 	local name="$1"
 	local dir="$2"
 	local conf="$3"
-
-	local script_security
-
-	config_get script_security "$name" script_security
-	[ -z "$script_security" ] && {
-		script_security=2
-	}
+	local security="$4"
 
 	procd_open_instance "$name"
 	procd_set_param command "$PROG"	\
@@ -89,7 +83,7 @@ openvpn_add_instance() {
 		--config "$conf" \
 		--up "/usr/lib/openvpn/openvpn-hotplug up $name" \
 		--down "/usr/lib/openvpn/openvpn-hotplug down $name" \
-		--script-security "$script_security"
+		--script-security "${security:-2}"
 	procd_set_param file "$dir/$conf"
 	procd_set_param term_timeout 15
 	procd_set_param respawn
@@ -110,11 +104,14 @@ start_instance() {
 		return 1
 	}
 
+	local script_security
+	config_get script_security "$s" script_security
+
 	[ ! -d "/var/run" ] && mkdir -p "/var/run"
 
 	if [ ! -z "$config" ]; then
 		append UCI_STARTED "$config" "$LIST_SEP"
-		openvpn_add_instance "$s" "${config%/*}" "$config"
+		openvpn_add_instance "$s" "${config%/*}" "$config" "$script_security"
 		return
 	fi
 
@@ -125,7 +122,7 @@ start_instance() {
 	append_params "$s" $OPENVPN_PARAMS
 	append_list "$s" $OPENVPN_LIST
 
-	openvpn_add_instance "$s" "/var/etc" "openvpn-$s.conf"
+	openvpn_add_instance "$s" "/var/etc" "openvpn-$s.conf" "$script_security"
 }
 
 start_service() {
diff --git a/package/network/services/openvpn/files/usr/lib/openvpn/openvpn-hotplug b/package/network/services/openvpn/files/usr/lib/openvpn/openvpn-hotplug
index f0c4d28810..9235fbacfe 100644
--- a/package/network/services/openvpn/files/usr/lib/openvpn/openvpn-hotplug
+++ b/package/network/services/openvpn/files/usr/lib/openvpn/openvpn-hotplug
@@ -5,4 +5,6 @@ shift
 INSTANCE=$1
 shift
 
-env -i ACTION=$ACTION INSTANCE=$INSTANCE /sbin/hotplug-call openvpn "$*"
+export ACTION=$ACTION
+export INSTANCE=$INSTANCE
+exec /sbin/hotplug-call openvpn "$@"
```

**Comment by feckert** [comment]
@jow- thanks for your feedback.

```
 +get_option() {
 +	local variable="$1"
 +	local option="$2"
 +
 +	local value="$(sed -rne 's/^[ \t]*'"$option"'[ \t]+([^ \t]+)[ \t]*$/\1/p' "$config" | tail -n1)"
 +	[ -n "$value" ] || value="$(sed -rne 's/^[ \t]*'"$option"'[ \t]+'"'([^']+)'"'[ \t]*$/\1/p' "$config" | tail -n1)"
 +	[ -n "$value" ] || value="$(sed -rne 's/^[ \t]*'"$option"'[ \t]+"(([^"\\]|\\.)+)"[ \t]*$/\1/p' "$config" | tail -n1 | sed -re 's/\\(.)/\1/g')"
 +	[ -n "$value" ] || return 1
 +
 +	export -n "$variable=$value"
 +	return 0
 +}
 +
```
Since I do not know your usecase exactly, I am not quite sure what you are doing in this function. Is the uci parsed there or is the openvpn config parsed there. I think the second?

Definitely Impressive sed command :+1: !

Updated my pullrquest with your requested changes.
Many thanks for this.

**Comment by jow-** [comment]
This function extracts a given parameter from the native OpenVPN configuration (stored in the `$config` environment variable set by the OpenVPN process when invoking `--up` and `--down` scripts.

The three different sed scripts are needed because the OpenVPN config parser accepts three different quoting styles for options:
 - unquoted: the option value must not contain spaces or the spaces must be escaped with a backslash and backslashes must be escaped with backslashes as well (e.g. `option up c:\\foo\ bar` -> `c:\foo bar`)
 - single quoted: the option value is enclosed in single quotes but may not contain any single quotes or escape sequences. Backslashes are treated literally (e.g. `option up 'c:\foo\ bar'` -> `c:\foo\ bar`)
 - double quoted: the option value is enclosed in double quotes, enclosed quotes or backslashes must be escaped with backslashes (e.g. `option up "date +\"%s\" --"` -> `date +"%s"`)

I noticed that my suggested code above mishandles the first case. The fix on top of your current PR is:

```diff
diff --git a/package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user b/package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user
index 41b9265001..6d45f0b7c6 100644
--- a/package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user
+++ b/package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user
@@ -4,7 +4,7 @@ get_option() {
 	local variable="$1"
 	local option="$2"
 
-	local value="$(sed -rne 's/^[ \t]*'"$option"'[ \t]+([^ \t]+)[ \t]*$/\1/p' "$config" | tail -n1)"
+	local value="$(sed -rne 's/^[ \t]*'"$option"'[ \t]+(([^ \t\\]|\\.)+)[ \t]*$/\1/p' "$config" | tail -n1 | sed -re 's/\\(.)/\1/g')"
 	[ -n "$value" ] || value="$(sed -rne 's/^[ \t]*'"$option"'[ \t]+'"'([^']+)'"'[ \t]*$/\1/p' "$config" | tail -n1)"
 	[ -n "$value" ] || value="$(sed -rne 's/^[ \t]*'"$option"'[ \t]+"(([^"\\]|\\.)+)"[ \t]*$/\1/p' "$config" | tail -n1 | sed -re 's/\\(.)/\1/g')"
 	[ -n "$value" ] || return 1
```

**Comment by feckert** [comment]
Thank you for your clarifying words.
Updated my pull request with your suggested change.
@jow- done

**Comment by jow-** [comment]
Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/jow.git

**Comment by ianchi** [comment]
This changed the order of the parameters sent to the old `up` script and broke it.
In may particular case it made the router completely unreachable.

**Comment by feckert** [comment]
@ianchi What was the order before that change?

**Comment by ianchi** [comment]
Basically now all arguments are shifted 3 positions to the right.
This are some examples. My script was originally called with no fixed arguments in the `up/down` option, just the script name.

### Original
```
DEV=$1
LOCAL_IP=$4
REMOTE_IP=$5
mode=$6
```

### New
```
DEV=$4
LOCAL_IP=$7
REMOTE_IP=$8
mode=$9
```

The three new arguments are:
`$ACTION` `$INSTANCE` `'openvpn'` 

I think in the case of the *legacy* scripts, this three arguments should be removed before the call to keep OpenVpn's native arguments:

in  package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user
```
# Wrap user defined scripts on up/down events
case "$ACTION" in
	up|down)
		if get_option command "$ACTION"; then
-			exec /bin/sh -c "$command $ACTION $INSTANCE $*"
+			shift
+			exec /bin/sh -c "$command $*"
		fi
	;;
esac
```

**Comment by drakesoft** [comment]
I have a big problem. Because openvpn does not execute my up script. I think this is caused by this hotplug.d change. I tired to move my script to /etc/openvpn.user but the variables ${dev} and  ${route_vpn_gateway} are not set. Without this variables my script wont work. How is the new way to get this variable?

**Comment by drakesoft** [comment]
I found a workaround. It seems the get_option function is not working because up and down script is not placed in the parsed config. So I changed it to read the up/down script variable with uci. 
Here is my modification of /etc/hotplug.d/openvpn/01-user ...

```
#!/bin/sh

[ -e "/etc/openvpn.user" ] && {
        env -i ACTION="$ACTION" INSTANCE="$INSTANCE" \
                /bin/sh \
                /etc/openvpn.user \
                $*
}

# Wrap user defined scripts on up/down events
case "$ACTION" in
        up|down)
                command=$(uci get openvpn.$INSTANCE.$ACTION)
                if  [ ! -z "$command" ] 
                then
                        shift
                        exec /bin/sh -c "$command $*"
                fi
        ;;
esac

exit 0

```

**Comment by feckert** [comment]
@jow- I currently have no use or setup for the problem and there is a lot of sed magic in the get_openvpn_options function. Can you look at it again?

**Comment by drakesoft** [comment]
I guess function get_openvpn_options parse the config file for openvpn --config parameter which does not contain the path of the up/down command because it is set via commandline to "/usr/libexec/openvpn-hotplug". So it simply never find the path to the custom up/down script.
